api/src/main/java/org/openmrs/api/impl/DiagnosisServiceImpl.java

```diff
@@ -0,0 +1,170 @@
+/**
+ * This Source Code Form is subject to the terms of the Mozilla Public License,
+ * v. 2.0. If a copy of the MPL was not distributed with this file, You can
+ * obtain one at http://mozilla.org/MPL/2.0/. OpenMRS is also distributed under
+ * the terms of the Healthcare Disclaimer located at http://openmrs.org/license.
+ *
+ * Copyright (C) OpenMRS Inc. OpenMRS is a registered trademark and the OpenMRS
+ * graphic logo is a trademark of OpenMRS Inc.
+ */
+package org.openmrs.api.impl;
+
+import java.util.ArrayList;
+import java.util.Date;
+import java.util.HashSet;
+import java.util.Iterator;
+import java.util.List;
+import java.util.Set;
+
+import org.openmrs.CodedOrFreeText;
+import org.openmrs.Diagnosis;
+import org.openmrs.Encounter;
+import org.openmrs.Patient;
+import org.openmrs.api.APIException;
+import org.openmrs.api.DiagnosisService;
+import org.openmrs.api.context.Context;
+import org.openmrs.api.db.DiagnosisDAO;
+import org.openmrs.parameter.EncounterSearchCriteria;
+import org.openmrs.parameter.EncounterSearchCriteriaBuilder;
+import org.openmrs.util.PrivilegeConstants;
+
+/**
+ * API functionality for Diagnoses.
+ *
+ * @since 2.2
+ */
+public class DiagnosisServiceImpl extends BaseOpenmrsService implements DiagnosisService {
+
+	public static final String DIAGNOSIS_ORDER_PRIMARY = "Primary";
+
+	private DiagnosisDAO dao;
+
+	public void setDiagnosisDAO(DiagnosisDAO dao) {
+		this.dao = dao;
+	}
+
+	/**
+	 * @see org.openmrs.api.DiagnosisService#save(Diagnosis)
+	 */
+	public Diagnosis save(Diagnosis diagnosis) {
+		if (diagnosis.getId() == null) {
+			Context.requirePrivilege(PrivilegeConstants.ADD_DIAGNOSES);
+		} else {
+			Context.requirePrivilege(PrivilegeConstants.EDIT_DIAGNOSES);
+		}
+		return dao.saveDiagnosis(diagnosis);
+	}
+
+	/**
+	 * @see org.openmrs.api.DiagnosisService#voidDiagnosis(Diagnosis, String)
+	 */
+	public Diagnosis voidDiagnosis(Diagnosis diagnosis, String voidReason) {
+		diagnosis.setVoided(true);
+		diagnosis.setVoidedBy(Context.getAuthenticatedUser());
+		diagnosis.setVoidReason(voidReason);
+		diagnosis.setDateVoided(new Date());
+		return save(diagnosis);
```

Comment:
You shouldn't be calling save directly because you will skip any logic around the spring proxy object.
You need to replace this with:

`Context.getDiagnosisService().save(diagnosis);`

Reply:
@wluyima I have changed this as recommended.